### PR TITLE
[NeuralChat] Enable retrieval with url as inputs

### PR DESCRIPTION
**Pull request by XuhuiRen**
add the support for crawling the data from a given list of URL link

### CODE REVIEW

File: intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/html_parser.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
+import requests
```
Comment (hshen14) [inline intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/html_parser.py:1]: add header for license

File: intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/html_parser.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+urllib3.disable_warnings()
+
+class Crawler:
+    def __init__(self, pool=None):
+        if pool:
+            assert isinstance(pool, (str, list, tuple)), 'url pool should be str, list or tuple'
+        self.pool = pool
+        self.headers = {
+            'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7',
```
Comment (hshen14) [inline intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/html_parser.py:21]: is it browser specific?

**Review by hshen14 — COMMENTED** [review]
Add UTs

File: intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/html_parser.py (hunk 2 above)
Comment (XuhuiRen) [reply intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/html_parser.py:38]: just confirmed with Guo Heng, this is just a simulator to imitate the behavior of a browser

File: intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/html_parser.py (hunk 1 above)
Comment (XuhuiRen) [reply intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/html_parser.py:18]: added

**Comment by XuhuiRen** [comment]
> Add UTs

added

File: intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/indexing.py [hunk 3]
```diff
@@ -115,110 +140,99 @@ def KB_construct(self, input):
...
+                                             persist_directory=self.persist_dir)
+            vectordb.persist()
+            print("The local knowledge base has been successfully built!")
+            return vectordb
+        elif self.retrieval_type == "sparse":
+            if self.document_store == "inmemory":
+                document_store = InMemoryDocumentStore(use_gpu=False, use_bm25=True)
+            elif self.document_store == "Elasticsearch":
+                document_store = ElasticsearchDocumentStore(host="localhost", index=self.index_name,
+                                                            port=9200, search_fields=["content", "title"])
```
Comment (hshen14) [inline intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/indexing.py:177]: shall we check whether port 9200 is available?

File: intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/indexing.py [hunk 4]
```diff
@@ -115,110 +140,99 @@ def KB_construct(self, input):
...
+                data_collection = self.parse_html(input)
+            except:
+                print("The given link/str is unavailable. Please try another one!")
+        else:
+            print("The input format is invalid!")
 
+        if self.retrieval_type == "dense":
+            documents = []
+            for data, meta in data_collection:
+                if len(data) < 5:
```
Comment (hshen14) [inline intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/indexing.py:210]: what does the magic number mean?

File: intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/indexing.py [hunk 5]
```diff
@@ -115,110 +140,99 @@ def KB_construct(self, input):
...
+            print("The local knowledge base has been successfully built!")
+            return Chroma(persist_directory=self.persist_dir, embedding_function=self.embeddings)
+        elif self.retrieval_type == "sparse":
+            if self.document_store == "Elasticsearch":
+                document_store = ElasticsearchDocumentStore(host="localhost", index=self.index_name,
+                                                            port=9200, search_fields=["content", "title"])
                 documents = []
                 for data, meta in data_collection:
+                    metadata = {"source": meta}
                     if len(data) < 5:
```
Comment (hshen14) [inline intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/indexing.py:228]: same as above

File: intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/retrieval_agent.py [hunk 6]
```diff
@@ -16,6 +16,7 @@
 # limitations under the License.
 
 import os
+from typing import Any, Dict, Iterator, List, Optional, Union, cast
```
Comment (hshen14) [inline intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/retrieval_agent.py:19]: do we really use all the types?

File: intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/indexing.py (hunk 5 above)
Comment (XuhuiRen) [reply intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/indexing.py:228]: meta data here means the reference link/document_path of the data, we set a length filter of 5 to filter some noisy data whose length is to short. In the previous products the customer may provide lots of too short and no-means data to us

File: intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/indexing.py (hunk 3 above)
Comment (XuhuiRen) [reply intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/indexing.py:177]: this part is transferring from haystack to langchain. This part would be removed in yuxiang's PR

File: intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/indexing.py (hunk 5 above)
Comment (hshen14) [reply intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/indexing.py:228]: define a global meaningful variable and refer to that

Comment (XuhuiRen) [reply intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/indexing.py:228]: > define a global meaningful variable and refer to that

will be fixed with the new vector database/retrieval/indexing change

File: intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/retrieval_agent.py (hunk 6 above)
Comment (XuhuiRen) [reply intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/retrieval_agent.py:19]: removed unnecessary import

**Pull request merged**